### PR TITLE
Add typehints and docstrings everywhere

### LINKED ISSUE #37: Make typehints consistent everywhere

### LINKED ISSUE #38: Add docstrings everywhere

### PR DESCRIPTION
**Pull request by nvaytet**
Fixes #37 
Fixes #38

### CODE REVIEW

File: src/plopp/graphics/basefig.py [hunk 1]
```diff
@@ -5,28 +5,51 @@
...
+        and update the relevant artist.
+
+        Parameters
+        ----------
+        *message:
+            The notification message containing the node id it originated from.
+        """
         node_id = message["node_id"]
         new_values = self.graph_nodes[node_id].request_data()
         self.update(new_values=new_values, key=node_id)
```
Comment (jl-wynen) [inline src/plopp/graphics/basefig.py:38]: Missing argument `draw`.
Or the abstractmethod shouldn't have that arg.

File: src/plopp/graphics/canvas3d.py [hunk 2]
```diff
@@ -5,13 +5,29 @@
...
+
+    Parameters
+    ----------
+    figsize:
+        The width and height of the renderer in pixels.
+    title:
+        The title to be placed above the figure.
+    """
+
+    def __init__(self, figsize: Tuple[float, float] = None, title: str = None):
```
Comment (jl-wynen) [inline src/plopp/graphics/canvas3d.py:30]: Doesn't call super's `__init__`, intentional?

File: src/plopp/graphics/canvas3d.py [hunk 3]
```diff
@@ -5,13 +5,29 @@
...
+    It is based on the ``pythreejs`` library. It provides a scene, camera, controls,
+    as well as functions for controlling the camera.
+    In addition, it draws a RGB axes helper for the XYZ dimensions, and can draw an
+    outline box, given a spatial range in the three directions.
+
+    Parameters
+    ----------
+    figsize:
+        The width and height of the renderer in pixels.
+    title:
```
Comment (jl-wynen) [inline src/plopp/graphics/canvas3d.py:26]: Unused?

File: src/plopp/graphics/fig1d.py [hunk 4]
```diff
@@ -7,23 +7,61 @@
...
+        dimension to be cropped.
+    title:
+        The figure title.
+    figsize:
+        The width and height of the figure, in inches.
+    **kwargs:
+        All other kwargs are forwarded to Matplotlib:
+        - ``matplotlib.pyplot.plot`` for 1d data with a non bin-edge coordinate
+        - ``matplotlib.pyplot.step`` for 1d data with a bin-edge coordinate
+    """
```
Comment (jl-wynen) [inline src/plopp/graphics/fig1d.py:50]: Unused: vmin, vmax, title

File: src/plopp/graphics/fig3d.py [hunk 5]
```diff
@@ -1,25 +1,62 @@
...
+        a colormap with that single color will be used.
+    mask_cmap:
+        The name of the colormap for masked data.
+    norm:
+        Control the scaling on the vertical axis.
+    vmin:
+        Lower bound for the vertical axis.
+    vmax:
+        Upper bound for the vertical axis.
+    title:
```
Comment (jl-wynen) [inline src/plopp/graphics/fig3d.py:40]: unused

File: src/plopp/graphics/outline.py [hunk 6]
```diff
@@ -38,8 +53,26 @@ def _make_sprite(string, position, color="black", size=1.0):
...
+    tick_size:
+        A number to scale the tick size.
+    """
 
-    def __init__(self, limits, tick_size=None):
+    def __init__(self,
+                 limits: Tuple[Variable, Variable, Variable],
+                 tick_size: float = None):
 
         center = [var.mean().value for var in limits]
```
Comment (jl-wynen) [inline src/plopp/graphics/outline.py:77]: Using this below causes type error `Expected type 'Tuple[float, float, float]', got 'list' instead `

File: src/plopp/graphics/outline.py [hunk 7]
```diff
@@ -58,9 +91,10 @@ def __init__(self, limits, tick_size=None):
         for obj in (self.box, self.ticks, self.labels):
             self.add(obj)
 
-    def _make_ticks(self, limits, center, tick_size):
+    def _make_ticks(self, limits: Tuple[Variable, Variable, Variable],
+                    center: Tuple[float, float, float], tick_size: float) -> p3.Group:
```
Comment (jl-wynen) [inline src/plopp/graphics/outline.py:95]: `center` is unused

File: src/plopp/core/utils.py [hunk 8]
```diff
@@ -45,35 +65,50 @@ def repeat(x, dim, n):
                                                                to=dim)
 
 
-def number_to_variable(x):
+def number_to_variable(x: Union[int, float]) -> sc.Variable:
```
Comment (jl-wynen) [inline src/plopp/core/utils.py:68]:
```suggestion
def number_to_variable(x: Union[int, float, sc.Variable]) -> sc.Variable:
```

File: src/plopp/functions/common.py [hunk 9]
```diff
@@ -36,23 +44,50 @@ def _to_data_array(obj):
...
 
-def preprocess(obj, crop=None, name='', ignore_size=False):
+def preprocess(obj: Union[ndarray, Variable, DataArray],
+               crop: Dict[str, Dict[str, Variable]] = None,
+               name: str = '',
+               ignore_size: bool = False):
+    """
+    Pre-process input data for plotting.
+    This involves:
+      - converting the input to a data array
```
Comment (jl-wynen) [inline src/plopp/functions/common.py:76]: reSt requires an empty line before an item list

Didn't check for other places...

File: src/plopp/graphics/canvas3d.py (hunk 2 above)
Comment (nvaytet) [reply src/plopp/graphics/canvas3d.py:30]: Nice catch, this doesn't need to inherit from anything actually. Thanks!

File: src/plopp/graphics/canvas3d.py (hunk 3 above)
Comment (nvaytet) [reply src/plopp/graphics/canvas3d.py:26]: Yes, for now, I added a TODO comment. I basically need to add a text widget in the top bar, above the canvas

File: src/plopp/graphics/canvas3d.py (hunk 2 above)
Comment (jl-wynen) [reply src/plopp/graphics/canvas3d.py:30]: Thank PyCharm. I just followed looked at its warnings :)

File: src/plopp/graphics/fig1d.py (hunk 4 above)
Comment (nvaytet) [reply src/plopp/graphics/fig1d.py:50]: Indeed. i opened #62

**Pull request merged**